crates/utils/src/interval_map/mod.rs

```diff
@@ -0,0 +1,1007 @@
+    fn right_rotate(&mut self, x: NodeIndex<Ix>) {
+        if self.left_ref(x, Node::is_sentinel) {
+            return;
+        }
+        let y = self.node_ref(x, Node::left);
+        self.node_mut(x, Node::set_left(self.node_ref(y, Node::right)));
+        if !self.right_ref(y, Node::is_sentinel) {
+            self.right_mut(y, Node::set_parent(x));
+        }
+
+        self.replace_parent(x, y);
+        self.node_mut(y, Node::set_right(x));
+
+        self.rotate_update_max(x, y);
+    }
+
+    /// Replaces parent during a rotation.
+    fn replace_parent(&mut self, x: NodeIndex<Ix>, y: NodeIndex<Ix>) {
+        self.node_mut(y, Node::set_parent(self.node_ref(x, Node::parent)));
+        if self.parent_ref(x, Node::is_sentinel) {
+            self.root = y;
+        } else if self.is_left_child(x) {
+            self.parent_mut(x, Node::set_left(y));
+        } else {
+            self.parent_mut(x, Node::set_right(y));
+        }
+        self.node_mut(x, Node::set_parent(y));
+    }
+
+    /// Updates the max value after a rotation.
+    fn rotate_update_max(&mut self, x: NodeIndex<Ix>, y: NodeIndex<Ix>) {
+        self.node_mut(y, Node::set_max_index(self.node_ref(x, Node::max_index)));
+        self.recaculate_max(x);
+    }
+
+    /// Updates the max value towards the root
+    fn update_max_bottom_up(&mut self, x: NodeIndex<Ix>) {
+        let mut p = x;
+        while !self.node_ref(p, Node::is_sentinel) {
+            self.recaculate_max(p);
+            p = self.node_ref(p, Node::parent);
+        }
+    }
+
+    /// Recaculate max value from left and right childrens
+    fn recaculate_max(&mut self, x: NodeIndex<Ix>) {
+        self.node_mut(x, Node::set_max_index(x));
+        let x_left = self.node_ref(x, Node::left);
+        let x_right = self.node_ref(x, Node::right);
+        if self.max(x_left) > self.max(x) {
+            self.node_mut(
+                x,
+                Node::set_max_index(self.node_ref(x_left, Node::max_index)),
+            );
+        }
+        if self.max(x_right) > self.max(x) {
+            self.node_mut(
+                x,
+                Node::set_max_index(self.node_ref(x_right, Node::max_index)),
+            );
+        }
+    }
+
+    /// Finds the node with the minimum interval.
+    fn tree_minimum(&self, mut x: NodeIndex<Ix>) -> NodeIndex<Ix> {
+        while !self.left_ref(x, Node::is_sentinel) {
+            x = self.node_ref(x, Node::left);
+        }
+        x
+    }
+
+    /// Replaces one subtree as a child of its parent with another subtree.
+    fn transplant(&mut self, u: NodeIndex<Ix>, v: NodeIndex<Ix>) {
+        if self.parent_ref(u, Node::is_sentinel) {
+            self.root = v;
+        } else {
+            if self.is_left_child(u) {
+                self.parent_mut(u, Node::set_left(v));
+            } else {
+                self.parent_mut(u, Node::set_right(v));
+            }
+            self.update_max_bottom_up(self.node_ref(u, Node::parent));
```

Comment:
Can we move `update_max_bottom_up` outside this function? Sometimes it can be called in batch, not necessary for every time.

Reply:
done